scapy/contrib/gtp.py

```diff
@@ -258,7 +258,7 @@ class GTPHeader(Packet):
     def post_build(self, p, pay):
         p += pay
         if self.length is None:
-            tmp_len = len(p) - 8
+            tmp_len = len(p) - 4 if self.version == 2 else len(p) - 8
```

Comment:
Could you explain how a test on the version fixes your issue?

Reply:
Absolutely! That field is defined differently for GTPv1 Headers and GTPv2 Headers. See for instance https://en.wikipedia.org/wiki/GPRS_Tunnelling_Protocol#Header

For GTPv1:
> Message Length
a 16-bit field that indicates the length of the payload in bytes (**rest of the packet following the mandatory 8-byte GTP header**). Includes the optional fields.

For GTPv2:
> Message length
This field shall indicate the length of the message in octets **excluding the mandatory of the GTP-C header (the first 4 octets)**.

So for GTPv1 you need to disregard 8 bytes, whereas for GTPv2 it's only 4. The GTPHeader class was correctly implementing the length calculation for GTPv1, but not for GTPv2.

I could also change the PR and overwrite this entire method in the gtp_v2.GTPHeader, which so far inherits all the logic from the gtp.GTPHeader: https://github.com/secdev/scapy/blob/master/scapy/contrib/gtp_v2.py#L242-L264
Now that I write this, I think that this is a much better place for this logic...